<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Steven J. Mack, PhD – steven.mack@ucsf.edu" />

<meta name="date" content="2020-04-10" />

<title>pould: Phased Or Unphased Linkage Disequilibrium</title>






<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">pould: Phased Or Unphased Linkage Disequilibrium</h1>
<h4 class="author">Steven J. Mack, PhD – <a href="mailto:steven.mack@ucsf.edu" class="email">steven.mack@ucsf.edu</a></h4>
<h4 class="date">April 10, 2020</h4>



<ul>
<li>Package Version: 1.0.0</li>
</ul>
<div id="overview" class="section level1">
<h1>Overview</h1>
<p>The <em>pould</em> package (pronounced “pooled”) calculates four linkage disequilibrium (LD) statistics – <em>D<sup>’</sup></em>, <em>W<sub>n</sub></em> and the conditional asymmetric LD (cALD) measures, <em>W<sub>A/B</sub></em> and <em>W<sub>B/A</sub></em>, for genotype data from pairs of genetic loci, and can treat these data as either phased or unphased for these calculations. The package includes a wrapper function that parses either column-formatted genotypes or multi-locus haplotypes in the <a href="http://17ihiw.org">17th International HLA and Immunogenetics Workshop</a> (IHIW) Family Haplotype Project’s HaplObserve output format. This wrapper function generates output files in a user-defined directory.</p>
<p>The package includes a function that applies a sign test to LD values for phased and unphased haplotypes generated by the wrapper function for haplotype- or genotype-formatted datasets, and a function that generates heat-maps for each LD measure. For more information, see:<br> Osoegawa et al. <a href="https://doi.org/10.1016/j.humimm.2019.01.010">Hum Immunol. 2019;80(9):633-643</a>.<br> Osoegawa et al. <a href="https://doi.org/10.1016/j.humimm.2019.05.018">Hum Immunol. 2019;80(9):644-660</a>.</p>
<p>For information about these LD measures, see:<br> Hedrick PW. <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1203208/pdf/331.pdf">Genetics; 1987,117,331-41</a>.<br> Cramér H. Mathematical Models of Statistics, 1946, Princeton University Press, Princeton NJ.<br> Thomson G, Single RM. <a href="https://doi.org/10.1534/genetics.114.165266">Genetics. 2014;198(1):321-31</a>.</p>
<p><em>Pould</em> accepts genotype and haplotype data for <em>individual subjects</em> as input data. To calculate the cALD measures using haplotype frequency data, try the <a href="https://CRAN.R-project.org/package=asymLD"><em>asymLD</em></a> package. See: Single et al. <a href="https://doi.org/10.1016/j.humimm.2015.09.001">Hum Immunol. 2016;77(3):288-294</a> for more information about <em>asymLD</em>.</p>
<div id="functions-and-input-formats" class="section level2">
<h2>Functions and Input Formats</h2>
<div id="cald" class="section level3">
<h3>cALD</h3>
<p>The <em>cALD()</em> function calculates the LD measures from genotype data for pairs of loci. The <code>drb1.dqb1.demo</code> dataset, shown below, illustrates the input format. The function accepts a 4-column data frame or tab-delimited text file with data for the first locus in columns 1 and 2, and data for the second locus in columns 3 and 4.</p>
<pre><code>&gt; drb1.dqb1.demo[1:6,]

##    DRB1  DRB1  DQB1  DQB1
## 1 15:01 07:01 06:02 03:03
## 2 04:05 13:01 03:02 06:03
## 3 04:01 13:02 03:02 06:09
## 4 16:02 04:05 05:02 03:02
## 5 15:01 07:01 06:02 02:02
## 6 04:01 04:01 03:02 03:02</code></pre>
<p>The locus names should be used as column headers, with one allele/variant in each column. Each row represents a single subject. The headers for columns 1 and 2 must be identical, as must the headers for columns 3 and 4. If phase is known, <em>cALD()</em> assumes that columns 1 and 3 represent one haplotype, and that columns 2 and 4 represent the second haplotype.</p>
<p>While HLA data are shown above, <em>cALD()</em> can accept any genetic data. The example below combines data for HLA-DQA1 and rs7743506.</p>
<pre><code>      HLA-DQA1       HLA-DQA1     rs7743506     rs7743506
        02:01          01:02           C             C
        04:01          01:02           A             C
        04:01          05:01           A             C
        01:01          01:02           C             C
        04:01          01:01           A             C
        05:01          01:02           C             C</code></pre>
</div>
<div id="ldwrap" class="section level3">
<h3>LDWrap</h3>
<p>The <em>LDWrap()</em> function parses either genotype data formatted in a two-column per locus format, or haplotype data formatted using the 17th International HLA and Immunogenetic Workshop Family Haplotype Project’s GL String-based format. The function accepts a data frame, a tab-formatted (.txt or .tsv) columnar genotype file, or a comma-separated value formatted (.csv) GL String haplotype file, and passes genotype data for all pairs of loci in that dataset to <em>cALD()</em> for LD analysis.</p>
<div id="input-formats" class="section level4">
<h4>Input Formats</h4>
<div id="haplotype-data" class="section level5">
<h5>Haplotype Data</h5>
<p>A minimal <em>LDWrap()</em> haplotype data file or data frame contains two columns named “Relation” and “Gl String”. Other columns are allowed, but are ignored. The <code>hla.hap.demo</code> dataset (shown below in edited form) illustrates the input format. Each row contains data for a single subject. The “Relation” column can contain any text string; however, values such as “mother”, “father” and “child” are standard for the Family HLA Data Project. <em>LDWrap()</em> will ignore all rows in which <code>Relation=child</code>; rows with any other value in the “Relation” column will be processed.</p>
<pre><code>Relation    Gl String
Subject HLA-A*02:01~HLA-C*07:02~HLA-B*07:02+HLA-A*01:01~HLA-C*06:02~HLA-B*57:01
Subject HLA-A*03:01~HLA-C*07:01~HLA-B*49:01+HLA-A*01:01~HLA-C*07:01~HLA-B*08:01
Subject HLA-A*11:01~HLA-C*04:01~HLA-B*15:01+HLA-A*03:01~HLA-C*08:02~HLA-B*14:02
Subject HLA-A*68:01~HLA-C*15:02~HLA-B*40:06+HLA-A*68:01~HLA-C*06:02~HLA-B*45:01</code></pre>
<p>The “Gl String” column contains a <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3715123/pdf/tan0082-0106.pdf">GL String</a> formatted multi-locus HLA haplotype. In GL String format, the ~ operator denotes phase, and the + operator denotes copies of genes (in this case diploidy). While GL Strings can be used to describe ambiguous alleles and genotypes using the / and | operators, ambiguous data cannot be included in an <em>LDWrap()</em> data file. <em>LDWrap()</em> requires that alleles be described as LOCUS<code>*</code>VARIANT (e.g., <code>HLA-DRB1*01:01</code>); locus prefixes (e.g., <code>HLA-</code>) are not required, but if locus refixes are included, all loci must be described using the same prefix. Allele data described without a locus (e.g., <code>01:01</code>) are not allowed. Unusual allele names (<code>HLA-A*NULL</code>, <code>HLA-DRB1*NoMatch</code>, <code>HLA-DPB1*NT</code>) and truncated versions of allele names (<code>HLA-A*01</code>, <code>HLA-A*01:01</code>, <code>HLA-A*01:01:01</code>, etc.) will be analyzed as distinct alleles, and may skew analytic results. <em>LDWrap()</em> includes an option to truncate colon-delimited allele names to specific numbers of fields for analysis.</p>
</div>
<div id="genotype-data" class="section level5">
<h5>Genotype Data</h5>
<p>A minimal <em>LDWrap()</em> genotype data file or data frame contains two columns per locus, with one allele in each column, as for <em>cALD()</em>, but accommodating more than two loci. Columns for the the same locus must be adjacent and can have identical names, or can be suffixed with &quot;_1&quot; and &quot;_2“. Columns named “SampleID” and “Disease” are permitted, but not required. No other columns are allowed. Allele names in each column can include a locus name (e.g., locus*allele); if locus names are not included, the locus name in the header will be associated with each allele in that column.</p>
</div>
<div id="locus-order" class="section level5">
<h5>Locus Order</h5>
<p>The order in which the loci appear, either in columns or in the GL String haplotype, affects the identification of haplotype locus pairs for analysis. For example, if HLA loci are organized alphabetically, haplotypes of the HLA-B and HLA-C loci will be analyzed as B~C haplotypes; if they are organized by map order, those haplotypes will be analyzed as C~B haplotypes. B~C and C~B haplotypes will not be recognized as the same by <em>LD.heat.map()</em>. To avoid this, it is recommended to use the same organization of loci for all analyses, and to use map order for HLA or KIR loci.</p>
</div>
</div>
</div>
<div id="ld.sign.test" class="section level3">
<h3>LD.sign.test</h3>
<p>The <em>LD.sign.test()</em> function applies the R Stats Package’s <em>binom.test()</em> function to pairs of LD values (<em>D’</em>, <em>Wn</em>, <em>WLoc1/Loc2</em>, <em>WLoc2/Loc1</em>), as well as the number of haplotypes, for phased and unphased haplotypes. The *_LD_results.csv* files generated by <em>LDWrap()</em> are the input files for this function, and generally, <em>LDWrap()</em> must be used before this function can be applied. See the <em>LDWrap()</em> Outputs section below for an example.</p>
</div>
<div id="ld.heat.map" class="section level3">
<h3>LD.heat.map</h3>
<p>The <em>LD.heat.map()</em> function generates heat-map plots of the LD values generated for each LD measure (<em>D’</em>, <em>Wn</em>, <em>WLoc1/Loc2</em>, and <em>WLoc2/Loc1</em>) for phased and unphased haplotpes. If LD values for only phased or unphased haplotpes are available, half-matrix heat-maps will be generated. The *_LD_results.csv* files generated by <em>LDWrap()</em> are the input files for this function, and generally, <em>LDWrap()</em> must be used before this function can be applied. See the <em>LDWrap()</em> Outputs section below for an example.</p>
</div>
</div>
<div id="outputs" class="section level2">
<h2>Outputs</h2>
<div id="cald-1" class="section level3">
<h3>cALD()</h3>
<p>By default, <em>cALD()</em> operates in “verbose” mode, and will write five lines of output to the console describing the phase-status of the LD analysis (phased or unphased), the loci and number of haplotypes analyzed, and the four LD measures calculated, as shown below.</p>
<pre><code>&gt; cALD(drb1.dqb1.demo)
## Calculating D', Wn and conditional ALD for 53 unphased genotypes at the DRB1 and DQB1 loci.
## D' for DRB1~DQB1 haplotypes: 0.95892767844544 (0.9589) 
## Wn for DRB1~DQB1 haplotypes: 0.811250972337927 (0.8113) 
## Variation of DQB1 conditioned on DRB1 (WDQB1/DRB1) = 0.904035615838528 (0.904)
## Variation of DRB1 conditioned on DQB1 (WDRB1/DQB1) = 0.778712696009626 (0.7787)</code></pre>
<p>When <code>verbose=FALSE</code>, <em>cALD()</em> returns a vector of <em>D<sup>’</sup></em>, <em>W<sub>n</sub></em>, <em>W<sub>B/A</sub></em>, <em>W<sub>A/B</sub></em> and the number of haplotypes, as below.</p>
<pre><code>&gt; cALDres &lt;- cALD(drb1.dqb1.demo, verbose=FALSE)
&gt; cALDres
## [1] &quot;0.958463650196244&quot; &quot;0.811184752436694&quot; &quot;0.903300938910147&quot; &quot;0.778712697633606&quot; &quot;53&quot;</code></pre>
<p>In addition, when saveVector=TRUE, <em>cALD()</em> will write a text file, containing a vector of all haplotypes, their frequencies and counts for the analyzed locus pair, to a user-specified directory. Unless specified via the <code>vecDir</code> parameter, this file is written to the directory specified by <em>tempdir()</em>. This vector file also includes information on the dataset and phase status applied to the genotype data for the analysis. An example generated for the <code>drb1.dqb1.demo</code> dataset is shown below.</p>
<pre><code>&gt; cALD(drb1.dqb1.demo,saveVector = TRUE)</code></pre>
<p>Haplotype vector file contents:</p>
<pre><code>Dataset                                         Phase   DRB1~DQB1   Frequency           Count
DRB1~DQB1_haplotype_Vector_2018-05-02_16-53-12  FALSE   01:01~02:01 0                   0
DRB1~DQB1_haplotype_Vector_2018-05-02_16-53-12  FALSE   01:02~02:01 0                   0
DRB1~DQB1_haplotype_Vector_2018-05-02_16-53-12  FALSE   01:03~02:01 0                   0
DRB1~DQB1_haplotype_Vector_2018-05-02_16-53-12  FALSE   03:01~02:01 0.094272076372315   79
DRB1~DQB1_haplotype_Vector_2018-05-02_16-53-12  FALSE   04:01~02:01 0                   0
DRB1~DQB1_haplotype_Vector_2018-05-02_16-53-12  FALSE   04:02~02:01 0                   0
DRB1~DQB1_haplotype_Vector_2018-05-02_16-53-12  FALSE   04:03~02:01 0                   0
DRB1~DQB1_haplotype_Vector_2018-05-02_16-53-12  FALSE   04:04~02:01 0                   0
.
.
.</code></pre>
</div>
<div id="ldwrap-1" class="section level3">
<h3>LDWrap</h3>
<p><em>LDWrap()</em> sends data to <em>cALD()</em>, captures the vector of LD results returned by <em>cALD()</em>, directs <em>cALD()</em> to write vectors of haplotypes for each locus pair into a user-specified directory, and writes a single table of LD results for all locus pairs to that same directory. As a single haplotype vector file is written for each locus pair, <em>LDWrap()</em> directs <em>cALD()</em> to write n(n-1)/2 haplotype vector files, where n is the number of loci in a haplotype. The only information <em>LDWrap()</em> returns to the console is a notification that the analysis has completed (<code>LD Analysis Complete</code>) and notifications that the provided dataset is missing the required columns. If the user specifies no destination for these files, they are written to the directory specified by <em>tempdir()</em>.</p>
<p>When all locus pairs have been analyzed by <em>cALD()</em>, <em>LDWrap()</em> writes a six-column CSV file (<code>*LD_results.csv</code>) of aggregated LD result vectors collected from <em>cALD()</em> to the specified directory. The column headers in this file are Loc1~Loc2 (identifying the locus pair), <em>D<sup>’</sup></em>, <em>W<sub>n</sub></em>, <em>W<sub>B/A</sub></em>, <em>W<sub>A/B</sub></em> and N_Haplotypes. An example of this file is shown below.</p>
<pre><code>&gt; LDWrap(hla.hap.demo)
LD Analysis Complete</code></pre>
<p>LD results file contents:</p>
<pre><code>Loc1~Loc2,D',Wn,WLoc1/Loc2,WLoc2/Loc1,N_Haplotypes
A~C,0.469024805013898,0.362566555750013,0.366359427624652,0.384413960789992,191
A~B,0.540780240853345,0.446662593270748,0.36839918931955,0.471334711300434,241
A~DRB1,0.400002012804198,0.335434108343871,0.27413544158564,0.320399398449896,233
A~DRB3,Not Calculated,Subject Threshold=10,Complete subjects=8,.,
.
.
.</code></pre>
<p><em>LDWrap()</em> attemtps to peform these LD calculations for all pairs of loci in the <em>LDWrap()</em> datafile. If a haplotype dataset includes locus pairs for which the number of subjects is below the <code>threshold</code> value (see “Parameters”, below), the <code>*LD_results.csv</code> file will include rows for locus pairs for which no LD calculations were performed. As shown above, those rows contain data similar to that shown for the A~DRB3 haplotpe – <code>Not Calculated</code>, <code>Subject Threshold=10</code>, <code>Complete subjects=8</code>, <code>.</code>, and ’ ’.</p>
<p>As shown below, in cases where at least one locus in a pair is monomorphic, no LD calculations are performed and the pertinent rows in the <code>*LD_results.csv</code> file will contain, <code>Not Calculated</code>, <code>Subject Threshold=10</code>, <code>Complete subjects=0</code>, <code>&quot;locusName&quot; is monomorphic.</code>, and ’ ’.</p>
<pre><code>Loc1~Loc2,D',Wn,WLoc1/Loc2,WLoc2/Loc1,N_Haplotypes
B~DRB3,Not Calculated,Subject Threshold=10,Complete subjects=130,DRB3 is monomorphic.,
DRB3~DRB4,Not Calculated,Subject Threshold=10,Complete subjects=130,DRB3 is monomorphic. DRB4 is monomorphic.,
DRB3~DQA1,Not Calculated,Subject Threshold=10,Complete subjects=93,DRB3 is monomorphic.,</code></pre>
</div>
<div id="ld.sign.test-1" class="section level3">
<h3>LD.sign.test</h3>
<p>For each LD measure and the number of haplotypes for phased and unphased versions of the same genotype data, <em>LD.sign.test()</em> reports the p-value of the sign test, comparing the number of locus pairs for which the value of the measure is higher in unphased haplotypes than phased haplotypes to the number of locus pairs for which that value is lower or equal. The function also reports the total number of locus pairs evaluated, and the number of locus pairs with equal values. These data can be reported in three ways; as a returned data frame, as a table written to the console, and as a CSV file written to a user-specified directory. All three report formats are illustrated below. Note, only the significance of the sign test is reported; when a significant trend is indicated, the directionality of the trend is not reported.</p>
<pre><code>&gt; LD.res &lt;- LD.sign.test(&quot;hla-family-data&quot;)
&gt; View(LD.res)   </code></pre>
<p>Returned Data Frame</p>
<pre><code>                             D'           Wn   WLoc1/Loc2   WLoc2/Loc1 N_Haplotypes
#unphased &gt; phased 1.500000e+01 1.400000e+01 1.500000e+01 1.500000e+01 0.000000e+00
#unphased = phased 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00
#locus pairs       1.500000e+01 1.500000e+01 1.500000e+01 1.500000e+01 1.500000e+01
p-values           6.103516e-05 9.765625e-04 6.103516e-05 6.103516e-05 6.103516e-05
</code></pre>
<pre><code>&gt; LD.sign.test(&quot;hla-family-data&quot;, returnFrame = FALSE)</code></pre>
<p>Console Table</p>
<pre><code>Sign Test results for the hla-family-data dataset for 15 locus pairs.
Measure         #U &gt; P  #U = P    p-value   
D'              15      0         6.104e-05 
Wn              14      0         0.0009766 
WLoc1/Loc2      15      0         6.104e-05 
WLoc2/Loc1      15      0         6.104e-05 
# Haplotypes    0       0         6.104e-05 </code></pre>
<p>CSV File</p>
<pre><code>,D',Wn,WLoc1/Loc2,WLoc2/Loc1,N_Haplotypes
#unphased &gt; phased,15,14,15,15,0
#unphased = phased,0,0,0,0,0
#locus pairs,15,15,15,15,15
p-values,6.10351562500001e-05,0.0009765625,6.10351562500001e-05,6.10351562500001e-05,6.10351562500001e-05</code></pre>
</div>
<div id="ld.heat.map-1" class="section level3">
<h3>LD.heat.map</h3>
<p>The <em>LD.heat.map()</em> function generates heat-map plots for each LD measure, visualizing the range of LD values for each locus-pair analzed using <em>LDWrap()</em>. LD values for phased data are presented in the upper half of the plot matrix, while values for unphased data are presnted in the lower half of the plot matrix. Color (blue to white to red) or greyscale (dark grey to light grey) plots can be specified. When <code>writePlot=TRUE</code>, PNG-formatted LD plots will be written to a directory identified using the <code>writeDir</code> parameter; this parameter defaults to the directory specified by <em>tempdir()</em>. When the <code>dataName</code> parameter is provided, heat-map plots are written with the name <code>&lt;dataName&gt;_&lt;LD measure&gt;_heatmap.png</code>. When the <code>phasedData</code> and <code>unphasedData</code> parameters are provided, the heat-map plot for each LD measure is named <code>&lt;phasedData&gt;-&lt;unphasedData&gt;_&lt;LD measure&gt;_heatmap.png</code>. When no <em>LDwrap()</em> results are available, or when the loci differ between the datasets specified by <code>phasedData</code> and <code>unphasedData</code>, a notification is written to the console. There are no other outputs.</p>
<pre><code>&gt; LD.heat.map(&quot;family-data&quot;)</code></pre>
<p>Console Notification</p>
<pre><code>Files family-data_Phased_LD_results.csv and family-data_Unphased_LD_results.csv are not found.</code></pre>
<pre><code>&gt; LD.heat.map(phasedData=&quot;my-data_phased.csv&quot;,unphasedData=&quot;my-data_unphased.csv&quot;)</code></pre>
<p>Console Notification</p>
<pre><code>The specified phased and unphased datafiles are not found.</code></pre>
<pre><code>&gt; LD.heat.map(phasedData=&quot;my-ABC-data_phased.csv&quot;,unphasedData=&quot;my-ABDRB-data-unphased.csv&quot;)</code></pre>
<p>Console Notification</p>
<pre><code>Different loci included in my-ABC-data_phased.csv and my-ABDRB-data-unphased.csv. Cannot produce heatmaps.</code></pre>
</div>
</div>
<div id="parameters" class="section level2">
<h2>Parameters</h2>
<div id="cald-2" class="section level3">
<h3>cALD</h3>
<pre><code>cALD(dataSet, inPhase = FALSE, verbose = TRUE, saveVector = FALSE, vectorName = &quot;&quot;, vectorPrefix = &quot;&quot;, vecDir = tempdir())</code></pre>
<p><strong>dataSet</strong></p>
<p>Class: Character. Required. No Default.</p>
<p>e.g., <code>dataSet=&quot;foo.txt&quot;</code> or <code>dataSet=drb1.dqb1.demo</code></p>
<p>Identifies the genotype data to be analyzed. <code>dataSet</code> should identify a four-column data-frame or tab-delimited text file. Columns 1 and 2 contain genotype data for one genetic locus, with one allele in each column. Columns 3 and 4 contain genotype data for the second locus. The headers for columns 1 and 2 identify the first locus, and must be identical. Similarly, the headers for columns 3 and 4 identify the second locus, and must be identical.</p>
<p><strong>inPhase</strong></p>
<p>Class: Logical. Required. Default=FALSE.</p>
<p>Identifies how the genotype data should be analyzed. When <code>inPhase=FALSE</code>, the expectation-maximization (EM) algorithm is applied to the genotype data to generate estimated haplotypes. LD values are calculated for those EM haplotypes. When <code>inPhase=TRUE</code>, the genotype data are treated as phased, with the alleles in column 1 in phase with those in column 3, and the alleles in column 2 in phase with those in column 4. LD values are calculated for those phased haplotypes.</p>
<p><strong>verbose</strong></p>
<p>Class: Logical. Required. Default=TRUE.</p>
<p>Identifies how LD values should be reported. When <code>verbose=TRUE</code>, values for <em>D<sup>’</sup></em>, <em>W<sub>n</sub></em>, <em>W<sub>B/A</sub></em>, <em>W<sub>A/B</sub></em> and the number of haplotypes evaluated are written to the console in a human readable form. When <code>verbose=FALSE</code>, those values are reported in a vector of length 5 and mode of “character”.</p>
<p><strong>saveVector</strong></p>
<p>Class: Logical. Required. Default=FALSE.</p>
<p>Specifies if a file containing the haplotype vector for the analyzed locus pair should be exported. If <code>saveVector=FALSE</code> no vector is written. If <code>saveVector=TRUE</code> a tab-delimited text file consisting of five columns is written to the working directory. The file includes the columns “Dataset”, “Phase”, the name of the analyzed locus pair, “Frequency”, and “Count”; the latter two describe the frequency and count data for all possible haplptypes.</p>
<p><strong>vectorName</strong></p>
<p>Class: Character. Optional. No Default. (<code>saveVector=TRUE</code> specific).</p>
<p>Provides a name for the tab-delimited haplotype vector file written when <code>saveVector=TRUE</code>. When <code>vectorName=&quot;foo&quot;</code> and <code>saveVector=TRUE</code>, the name of the haplotype vector file will be “foo.txt”. When <code>vectorName=&quot;&quot;</code> and <code>saveVector=TRUE</code>, the name of the haplotype vector file will include the loci analyzed and a time-stamp, formatted as “Locus1~Locus2_haplotype_Vector_yyyy-MM-dd-HH-mm-ss.txt”.</p>
<p><strong>vectorPrefix</strong></p>
<p>Class: Character. Optional. No Default. (<code>saveVector=TRUE</code> specific).</p>
<p>Applies a prefix that includes the applied phase-status to the name of the tab-delimited haplotype vector file written when <code>saveVector=TRUE</code>, when <code>vectorName=&quot;&quot;</code>. If any text string is provided for vectorName, this parameter is ignored. E.g., when <code>saveVector=TRUE</code>, <code>vectorName=&quot;&quot;</code>, <code>vectorPrefix=&quot;foo&quot;</code> and <code>inPhase=FALSE</code>, the haplotype vector file will be named “foo_unphased_Locus1~Locus2_haplotype_Vector_yyyy-MM-dd-HH-mm-ss.txt”.</p>
<p>When the <code>LDWrap()</code> function directs <code>cALD()</code> to write files containing haplotype vectors for each locus pair, <code>LDWrap()</code> provides the file information to <code>cALD</code> via the <code>vectorPrefix</code> parameter. The resulting files will contain the name of the family haplotype dataset processed by <code>LDWrap()</code>, the phase-status, haplotype pair-name, and a timestamp. If a positive <code>trunc</code> parameter was provided to <code>LDWrap()</code>, the truncation level will appear in the names of these haplotype vector files.</p>
<p><strong>vecDir</strong></p>
<p>Class: Character. Optional. Default=tempdir(). (<code>saveVector=TRUE</code> specific).</p>
<p>Specifies the directory into which the haplotype vector file should be written.</p>
</div>
<div id="ldwrap-2" class="section level3">
<h3>LDWrap</h3>
<pre><code>LDWrap(famData, threshold = 10, phased = TRUE, frameName = &quot;hla-family-data&quot;)</code></pre>
<p><strong>famData</strong></p>
<p>Class: Character. Required. No Default.</p>
<p>e.g., <code>famData=&quot;foo.csv&quot;</code> or <code>famData=hla.hap.demo</code></p>
<p>Identifies the haplotype or genotype dataset to be analyzed. For haplotype data, <code>famData</code> should identify a data frame or CSV file. This dataset must inlcude columns with the headers “Relation” and “Gl String”. If either (or both) of these column headers is not found in the dataset, or if the data file is not a CSV file, <em>LDWrap()</em> will halt the analysis with a notification about the missing header(s). For genotype data, <code>famData</code> should identify a data frame or tab-delimited text file with two columns of allele data for each locus. See the Functions and Input Formats section above for additional details about these dataset formats.</p>
<p><strong>threshold</strong></p>
<p>Class: Numeric. Required. Default=10.</p>
<p>Identifies the minimum number of subjects with haplotype data for a given locus pair required for the analysis of that locus pair. Analysis for that locus pair is not performed if the threshold is not met, and the LD results file will identify the threshold value and the number of subjects with data for that locus pair. If <code>threshold</code> is set to less than 1, it is automatically set to 1.</p>
<p><strong>phased</strong></p>
<p>Class: Logical. Required. Default=TRUE.</p>
<p>Specifies whether the haplotype data should be treated as phased (<code>phased=TRUE</code>) or unphased (<code>phased=FALSE</code>) for analysis.</p>
<p><strong>frameName</strong></p>
<p>Class: Character. Optional. Default=“hla-family-data”.</p>
<p>Provides a name that will be included in the names of the result files if <code>famData</code> specifies a data frame. The value of <code>frameName</code> is passed to <em>cALD()</em> as the <code>vectorPrefix</code> parameter. If <code>famData</code> specifies a file, <code>frameName</code> is ignored.</p>
<p><strong>trunc</strong></p>
<p>Class: Numeric. Required. Default=0.</p>
<p>Specifies the number of fields to which colon-delimited allele names in <code>famdData</code> should be truncated. The default value of 0 indicates no truncation. A value higher than the number of fields in the supplied allele data will result in no truncation. When a positive value of <code>trunc</code> is provided, the names of the output files will include the specified truncation level.</p>
<p><strong>writeTo</strong></p>
<p>Class: Character. Optional. Default=tempdir().</p>
<p>Specifies the directory into which LDWrap() should write files.</p>
</div>
<div id="ld.sign.test-2" class="section level3">
<h3>LD.sign.test</h3>
<pre><code>LD.sign.test(dataName,verbose = TRUE,returnFrame = FALSE)</code></pre>
<p><strong>dataName</strong></p>
<p>Class: Character. Required. No Default.</p>
<p>e.g., <code>dataName=&quot;foo&quot;</code></p>
<p>The “base” name of the &quot;_LD_result.csv&quot; files generated by <em>LDWrap()</em>, with the &quot;_Phased_LD_results.csv&quot; or &quot;_Unphased_LD_results.csv&quot; suffixes removed. This corresponds to the value of the <em>LDWrap()</em> <code>frameName</code> parameter when the <em>LDWrap()</em> <code>famData</code> parameter does not specify a file; e.g., when specifying the &quot;_LD_results.csv&quot; files generated by <em>LDWrap()</em> for the <code>hla.hap.demo</code> data included with this package, <code>dataName=&quot;hla-family-data&quot;</code>. If the corresponding &quot;<dataName>_Phased_LD_results.csv&quot; or &quot;<dataName>_Unphased_LD_results.csv&quot; files are not found, the function will halt with a notification identifying the file(s) that are not found.</p>
<p><strong>verbose</strong></p>
<p>Class: Logical. Required. Default=TRUE.</p>
<p>Identifies if messages about function progress and results should be displayed in the console (verbose=TRUE) or not (verbose=FALSE). The default is verbose=TRUE. In addition to the table of results and messages about missing input files, messages regarding locus pairs for which no LD values are available, and about discrepancies between locus pairs with available data in the phased and unphased datasets are also suppressed with verbose=FALSE.</p>
<p><strong>returnFrame</strong></p>
<p>Class: Logical. Required. Default=TRUE.</p>
<p>Identifies if a data frame of results should be returned (returnFrame=TRUE). If <code>returnFrame=FALSE</code>, a CSV file of results named &quot;<dataName>_LD-sign-test_results.csv&quot; is written in the directory specified by the <code>resultDir</code> parameter.</p>
<p><strong>resultDir</strong></p>
<p>Class: Character. Optional. Default=tempdir().</p>
<p>Specifies the directory into which LD.sign.test() should write the CSV file of results.</p>
</div>
<div id="ld.heat.map-2" class="section level3">
<h3>LD.heat.map</h3>
<pre><code>LD.heat.map(dataName)</code></pre>
<p><strong>dataName</strong></p>
<p>Class: Character. Optional. Default=&quot;&quot;.</p>
<p>e.g., <code>dataName=&quot;foo&quot;</code></p>
<p>The “base” name of the &quot;_LD_result.csv&quot; files generated by <em>LDWrap()</em>, with the &quot;_Phased_LD_results.csv&quot; or &quot;_Unphased_LD_results.csv&quot; suffixes removed. This corresponds to the value of the <em>LDWrap()</em> <code>frameName</code> parameter when the <em>LDWrap()</em> <code>famData</code> parameter does not specify a file; e.g., when specifying the &quot;_LD_results.csv&quot; files generated by <em>LDWrap()</em> for the <code>hla.hap.demo</code> data included with this package, <code>dataName=&quot;hla-family-data&quot;</code>. If the corresponding &quot;<dataName>_Phased_LD_results.csv&quot; and &quot;<dataName>_Unphased_LD_results.csv&quot; files are not found, the function will halt with a notification identifying the files that are not found. If this parameter is omitted, <code>phasedData</code> and/or <code>unphasedData</code> must be provided.</p>
<p><strong>phasedData</strong></p>
<p>Class: Character. Optional. Default=&quot;&quot;.</p>
<p>The complete name of a file of phased LD results generated by <em>LDWrap()</em>. Provide this filename if no “base” name is provided for <code>dataName</code> and you want to generate heat-maps for a specific set of phased LD values.</p>
<p>e.g., <code>phasedData=&quot;phased-data.csv&quot;</code></p>
<p><strong>unphasedData</strong></p>
<p>Class: Character. Optional. Default=&quot;&quot;.</p>
<p>The complete name of a file of unphased LD results generated by <em>LDWrap()</em>. Provide this filename if no “base” name is provided for <code>dataName</code> and you want to generate heat-maps for a specific set of unphased LD values.</p>
<p>e.g., <code>phasedData=&quot;unphased-data.csv&quot;</code></p>
<p><strong>phasedLabel</strong></p>
<p>Class: Character. Required. Default=“Phased”.</p>
<p>e.g., <code>phasedLabel=&quot;Pedigree Phased&quot;</code></p>
<p>Specifies the label that should appear on the heat-map plots for the upper (presumed phased) half of the plot.</p>
<p><strong>unphasedLabel</strong></p>
<p>Class: Character. Required. Default=“EM-estimated”.</p>
<p>e.g., <code>unphasedLabel=&quot;EM Haplotypes&quot;</code></p>
<p>Specifies the label that should appear on the heat-map plots for the lower (presumed unphased) half of the plot.</p>
<p><strong>color</strong></p>
<p>Class: Logical. Required. Default=TRUE.</p>
<p>e.g., <code>color=FALSE</code></p>
<p>Identifies if the heat-map plots should be generated in color (<code>color=TRUE</code>) or greyscale (<code>color=FALSE</code>). The default is <code>color=TRUE</code>. Color heat-map plots will range from blue (low LD values) to white (LD values of 0.5) to red (high LD values). Greyscale heat-map plots will range from dark grey (Low LD values) to light grey (high LD values).</p>
<p><strong>writePlot</strong></p>
<p>Class: Logical. Required. Default=FALSE.</p>
<p>Identifies if the heat-map plots should be automatically saved after they are generated.</p>
<p><strong>writeDir</strong></p>
<p>Class: Character. Optional. Default=tempdir().</p>
<p>The directory into which the heat-map plots should be saved when <code>writePlot=TRUE</code>. The default is the directory specified by <em>tempdir()</em>.</p>
</div>
</div>
<div id="examples" class="section level2">
<h2>Examples</h2>
<p>###cALD()</p>
<pre><code># Analyzing the included HLA-DRB1 HLA-DQB1 genotype data and reporting results in the console
cALD(drb1.dqb1.demo)
# Alternatively returning a vector of LD results, with nothing reported in the console
LDvec &lt;- cALD(drb1.dqb1.demo,verbose=FALSE)
LDvec
# Enforcing phase between columns 1 and 3 and between columns 2 and 4 for analysis
cALD(drb1.dqb1.demo,inPhase=TRUE)
# Writing the haplotype vector to a file in the temporary directory that will 
# have the time-stamped name DRB1~DQB1_haplotype_Vector_yyyy-MM-dd_HH-mm-ss.txt
cALD(drb1.dqb1.demo,saveVector = TRUE)
# Writing the haplotype vector to a file named &quot;foo.txt&quot; in the temporary directory  
cALD(drb1.dqb1.demo,saveVector = TRUE,vectorName = &quot;foo&quot;)
# Writing the haplotype vector to a file in the temporary directory with the prefix &quot;foo_&quot;
cALD(drb1.dqb1.demo,saveVector = TRUE,vectorPrefix = &quot;foo&quot;)
# Writing the haplotype vector to a file in the working directory
cALD(drb1.dqb1.demo,saveVector = TRUE,vecDir = getwd())</code></pre>
<p>###LDWrap()</p>
<pre><code># Analyzing the included HLA haplotype data 
# This will create 15 haplotype vector files and one LD results file in the temporary directory
LDWrap(hla.hap.demo)
# Specifying the prefix &quot;foo_Phased&quot; for the LD results file, and &quot;foo_phased&quot; for the haplotype vector files
LDWrap(hla.hap.demo,frameName = &quot;foo&quot;)
# Truncating the alleles in hla.hap.demo to 1 field for analysis.
LDWrap(hla.hap.demo,frameName = &quot;foo&quot;, trunc=1)
# Analyzing the included HLA genotype data
LDWrap(drb1.dqb1.demo,frameName=&quot;hla-genotype-data&quot;)
# Writing the resulting files to the working directory
LDWrap(hla.hap.demo,writeTo = getwd())</code></pre>
<p>###LD.sign.test()</p>
<pre><code># Generating LDWrap() results files for the example data included with this package in the temporary directory
LDWrap(hla.hap.demo)
LDWrap(hla.hap.demo,phased=FALSE)
# Analyzing the results files generated by LDWrap(), with a CSV of the results written to the temporary directory.
LDdata &lt;- paste(tempdir(),&quot;hla-family-data&quot;,sep=.Platform$file.sep)
LD.sign.test(LDdata, returnFrame=FALSE)
# Returning only a data frame for the same analysis.
LD.res &lt;- LD.sign.test(LDdata,verbose=FALSE)
View(LD.res)
# Writing the CSV file to the working directory
LD.sign.test(LDdata,returnFrame = FALSE,resultDir = getwd())</code></pre>
<p>###LD.heat.map()</p>
<pre><code># Generating LDWrap() results files for the example data included with this package in the working directory
LDWrap(hla.hap.demo, writeTo=getwd())
LDWrap(hla.hap.demo,phased=FALSE, writeTo=getwd())
# Generating color heat-map plots based on the LD result files in the working directory
LD.heat.map(&quot;hla-family-data&quot;)
# Generating greyscale heat-map plots based on the LD result files in the working directory
LD.heat.map(&quot;hla-family-data&quot;,color=FALSE)
# Generating heat-map plots for phased data alone based on the LD result files in the working directory
LD.heat.map(phasedData=&quot;hla-family-data_Phased_LD_results.csv&quot;,unphasedLabel=&quot;&quot;)
# Writing color PNG-formatted heat-map plots to the working directory, using the LD files in the working directory
LD.heat.map(&quot;hla-family-data&quot;,writePlot = TRUE,writeDir = getwd())</code></pre>
</div>
<div id="section" class="section level2">
<h2></h2>
<p>End of vignette.</p>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
